tests/test_async_client.py

```diff
@@ -8,12 +8,17 @@
 pytestmark = [pytest.mark.anyio, pytest.mark.network]
 
 
-@pytest.mark.parametrize("qos", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE])
-async def test_publish_subscribe(qos: QoS) -> None:
+@pytest.mark.parametrize(
+    "qos_sub", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE]
+)
+@pytest.mark.parametrize(
+    "qos_pub", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE]
+)
+async def test_publish_subscribe(qos_sub: QoS, qos_pub: QoS) -> None:
```

Comment:
What do these additions actually test for? This updated test passes just fine without the other changes in this PR.

Reply:
Broker exercising. ;-)

I initially added this code to check whether it was the publish or the subscribe part that caused the problem I was seeing (ultimately fixed in #20).

I left it in because checking every possible QoS combination seems prudent, so that we don't hit any other issues along these lines.

Reply:
But we're not really checking anything new here, are we? When I originally wrote this test, it wasn't passing because the I/O implementation wasn't handling the different QoS levels correctly. If these new test combinations don't add anything useful to the test suite, they should be left out. We're not trying to to validate the brokers here.

Reply:
The original test didn't tell the broker which QoS level we want the messages to have, and the test suite should include that.

Alternately I could split this test in two: one that exercises every `publish` QoS level, and one that does the same with all QoS levels in `subscribe`. That'd result in running six testcases instead of nine, which seems like a somewhat marginal improvement, but it's your call.

Reply:
If the intent is to test if we client can limit the maximum QoS level in its subscription, then the test should at least check the QoS level of the received messages.

Reply:
OK, that's easy enough to add.